File: backend/app/api/v1/services.py
"""Tenant-scoped service catalog endpoints.

Base path: /api/v1/tenant/services
Guards: manage/services = admin+manager for writes; all staff can read.
"""

from __future__ import annotations

import uuid
from typing import Any

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.dependencies import get_current_admin_user, require_permission
from app.db.session import get_session
from app.models.admin_user import AdminUser
from app.schemas.services import (
    MilestoneStepResponse,
    ServiceCreateRequest,
    ServiceDetailResponse,
    ServiceListItem,
    ServiceListResponse,
    ServiceUpdateRequest,
)
from app.services import services as service_service

router = APIRouter(prefix="/tenant/services", tags=["services"])


def _get_tenant_id(user: AdminUser) -> uuid.UUID:
    if user.tenant_id is None:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="User must belong to a tenant",
        )
    return user.tenant_id


def _to_detail_response(service: Any) -> ServiceDetailResponse:
    return ServiceDetailResponse(
        id=service.id,
        name=service.name,
        description=service.description,
        default_price=service.default_price,
        is_active=service.is_active,
        step_count=len(service.milestone_steps),
        created_at=service.created_at,
        updated_at=service.updated_at,
        steps=[
            MilestoneStepResponse.model_validate(s) for s in service.milestone_steps
        ],
    )


# ═══════════════════════════════════════════════════════════════════════════
# CRUD
# ═══════════════════════════════════════════════════════════════════════════


@router.post(
    "/",
    response_model=ServiceDetailResponse,
    status_code=status.HTTP_201_CREATED,
)
async def create_service_endpoint(
    body: ServiceCreateRequest,
    session: AsyncSession = Depends(get_session),
    user: AdminUser = Depends(require_permission("manage", "services")),
) -> ServiceDetailResponse:
    """Create a new service with optional milestone step templates. Admin/Manager only."""
    tenant_id = _get_tenant_id(user)
    service = await service_service.create_service(
        session,
        tenant_id=tenant_id,
        name=body.name,
        description=body.description,
        default_price=body.default_price,
        is_active=True,
        steps=body.steps,
        actor_id=user.id,
    )
    return _to_detail_response(service)


@router.get("/", response_model=ServiceListResponse)
async def list_services_endpoint(
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=20, ge=1, le=100),
    is_active: bool | None = Query(default=None),
    q: str | None = Query(default=None, min_length=1),
    sort: str | None = Query(default=None),
    session: AsyncSession = Depends(get_session),
    user: AdminUser = Depends(get_current_admin_user),
) -> ServiceListResponse:
    """List services for a tenant. All staff can read."""
    tenant_id = _get_tenant_id(user)

    result = await service_service.list_services(
        session,
        tenant_id=tenant_id,
        page=page,
        page_size=page_size,
        is_active=is_active,
        q=q,
        sort=sort,
    )

    items = [ServiceListItem(**item) for item in result["items"]]
    return ServiceListResponse(
        items=items,
        total=result["total"],
        page=result["page"],
        page_size=result["page_size"],
    )


@router.get("/{service_id}", response_model=ServiceDetailResponse)
async def get_service_endpoint(
    service_id: str,
    session: AsyncSession = Depends(get_session),
    user: AdminUser = Depends(get_current_admin_user),
) -> ServiceDetailResponse:
    """Get service detail with milestone step templates. All staff can read."""
    tenant_id = _get_tenant_id(user)

    try:
        sid = uuid.UUID(service_id)
    except ValueError:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Service not found",
        ) from None

    service = await service_service.get_service_detail(
        session, tenant_id=tenant_id, service_id=sid
    )
    return _to_detail_response(service)


@router.patch("/{service_id}", response_model=ServiceDetailResponse)
async def update_service_endpoint(
    service_id: str,
    body: ServiceUpdateRequest,
    session: AsyncSession = Depends(get_session),
    user: AdminUser = Depends(require_permission("manage", "services")),
) -> ServiceDetailResponse:
    """Update service fields. If `steps` provided, replaces full set. Admin/Manager only."""
    tenant_id = _get_tenant_id(user)

    try:
        sid = uuid.UUID(service_id)
    except ValueError:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Service not found",
        ) from None

    updates = body.model_dump(exclude_unset=True)

    service = await service_service.update_service(
        session,
        tenant_id=tenant_id,
        service_id=sid,
        updates=updates,
        actor_id=user.id,
    )
    return _to_detail_response(service)


@router.delete(
    "/{service_id}",
    status_code=status.HTTP_204_NO_CONTENT,
)
async def delete_service_endpoint(
    service_id: str,
    session: AsyncSession = Depends(get_session),
    user: AdminUser = Depends(require_permission("manage", "services")),
) -> None:
    """Hard-delete a service. Cascades to step templates. Admin/Manager only."""
    tenant_id = _get_tenant_id(user)

    try:
        sid = uuid.UUID(service_id)
    except ValueError:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Service not found",
        ) from None

    await service_service.delete_service(
        session, tenant_id=tenant_id, service_id=sid, actor_id=user.id
    )

File: frontend/src/lib/api/services.js
import { apiFetch } from './client.js';

/**
 * Tenant service-catalog API endpoints.
 *
 * GET endpoints are accessible to all staff (admin, manager, employee).
 * Write endpoints require admin or manager; the server enforces and the UI
 * hides/disables write controls for employees.
 */

/**
 * @typedef {object} MilestoneStep
 * @property {string} [id] present on read, absent on create/replace
 * @property {string} name
 * @property {number} sequence_order
 * @property {number|null} expected_duration_days
 * @property {string|null} [description]
 * @property {string} [created_at]
 * @property {string} [updated_at]
 */

/**
 * @typedef {object} ServiceListItem
 * @property {string} id
 * @property {string} name
 * @property {string|null} description
 * @property {string|null} default_price decimal-as-string, e.g. "1500.00"
 * @property {boolean} is_active
 * @property {number} step_count
 * @property {string} created_at
 * @property {string} updated_at
 */

/**
 * @typedef {ServiceListItem & { steps: MilestoneStep[], in_use: boolean, project_count: number }} ServiceDetail
 */

/**
 * @param {typeof fetch} fetchFn
 * @param {string} token
 * @param {{ page?: number, page_size?: number, is_active?: boolean|string, q?: string, sort?: string }} [params]
 * @returns {Promise<{ items: ServiceListItem[], total: number, page: number, page_size: number }>}
 */
export function listServices(fetchFn, token, params = {}) {
	return apiFetch(fetchFn, '/tenant/services', { token, params });
}

/**
 * @param {typeof fetch} fetchFn
 * @param {string} token
 * @param {string} id
 * @returns {Promise<ServiceDetail>}
 */
export function getService(fetchFn, token, id) {
	return apiFetch(fetchFn, `/tenant/services/${encodeURIComponent(id)}`, { token });
}

/**
 * @param {typeof fetch} fetchFn
 * @param {string} token
 * @param {{
 *   name: string,
 *   description?: string|null,
 *   default_price?: number|string|null,
 *   is_active?: boolean,
 *   steps?: Array<{ name: string, sequence_order: number, expected_duration_days?: number|null, description?: string|null }>
 * }} body
 * @returns {Promise<ServiceDetail>}
 */
export function createService(fetchFn, token, body) {
	return apiFetch(fetchFn, '/tenant/services', { method: 'POST', token, body });
}

/**
 * PATCH is partial. When `steps` is provided it REPLACES the full set;
 * the server renumbers sequence_order on save.
 *
 * @param {typeof fetch} fetchFn
 * @param {string} token
 * @param {string} id
 * @param {Record<string, any>} body partial fields (any may be null to clear)
 * @returns {Promise<ServiceDetail>}
 */
export function updateService(fetchFn, token, id, body) {
	return apiFetch(fetchFn, `/tenant/services/${encodeURIComponent(id)}`, {
		method: 'PATCH',
		token,
		body
	});
}

/**
 * @param {typeof fetch} fetchFn
 * @param {string} token
 * @param {string} id
 * @returns {Promise<null>}
 */
export function deleteService(fetchFn, token, id) {
	return apiFetch(fetchFn, `/tenant/services/${encodeURIComponent(id)}`, {
		method: 'DELETE',
		token
	});
}
